jdaviz/configs/imviz/tests/test_wcs_utils.py

```diff
@@ -108,58 +108,44 @@ def test_simple_gwcs():
     assert not result[-1]
 
 
-@pytest.mark.remote_data
-@pytest.mark.filterwarnings(r"ignore::astropy.wcs.wcs.FITSFixedWarning")
-def test_non_wcs_layer_labels(imviz_helper):
-    # load a real image w/ WCS:
-    real_image_path = get_pkg_data_filename('tutorials/FITS-images/HorseHead.fits')
-    imviz_helper.load_data(real_image_path)
-
-    # load a WCS-only layer:
-    ndd = wcs_utils._get_rotated_nddata_from_label(
-        app=imviz_helper.app,
-        data_label=imviz_helper.app.data_collection[0].label,
-        rotation_angle=0*u.deg
-    )
-    imviz_helper.load_data(ndd)
-
-    # confirm that only the image is labeled:
-    assert len(imviz_helper.app.state.layer_icons) == 2
-
-    # confirm the WCS-only layer is logged:
-    viewer = imviz_helper.app.get_viewer('imviz-0')
-    assert len(viewer.state.wcs_only_layers) == 1
-
-    # load a second WCS-only layer:
-    ndd2 = wcs_utils._get_rotated_nddata_from_label(
-        app=imviz_helper.app,
-        data_label=imviz_helper.app.data_collection[0].label,
-        rotation_angle=45*u.deg
-    )
-    imviz_helper.load_data(ndd2)
-    assert len(imviz_helper.app.state.layer_icons) == 3
-    assert len(viewer.state.wcs_only_layers) == 2
-
-    wcs_only_refdata_icon = 'mdi-compass-outline'
-    wcs_only_not_refdata_icon = 'mdi-compass-off-outline'
-
-    for i, data in enumerate(imviz_helper.app.data_collection):
-        viewer = imviz_helper.app.get_viewer("imviz-0")
-
-        if i == 0:
-            # first entry is image data:
-            assert imviz_helper.app.state.layer_icons[data.label] == 'a'
-            assert viewer.state.reference_data.label == data.label
-        else:
-            # icon before setting as refdata:
-            before_icon = imviz_helper.app.state.layer_icons[data.label]
-
-            # set as refdata:
-            imviz_helper.app._change_reference_data(data.label)
-            assert viewer.state.reference_data.label == data.label
-
-            # icon after setting as refdata:
-            after_icon = imviz_helper.app.state.layer_icons[data.label]
-
-            assert before_icon == wcs_only_not_refdata_icon
-            assert after_icon == wcs_only_refdata_icon
+# TODO: Add more tests.
+class TestWCSOnly(BaseImviz_WCS_GWCS):
+
+    # TODO: Replace private API calls with more public ones when available.
+    def test_non_wcs_layer_labels(self):
+        self.imviz.link_data(link_type="wcs")
+        assert len(self.imviz.app.data_collection) == 3
+
+        # Load a WCS-only layer, bypassing normal labeling scheme.
+        ndd = wcs_utils._get_rotated_nddata_from_label(
+            app=self.imviz.app,
+            data_label="fits_wcs[DATA]",
+            rotation_angle=5 * u.deg
+        )
+        wcs_only_data_label = self.imviz.app._wcs_only_label
+        for wcs_only_data, _ in _nddata_to_glue_data(ndd, wcs_only_data_label):
+            break  # Can stop after first iteration

```

Comment:
I found it a bit confusing the first few times I read it, but I think it's what we need and it's neat so I'm not asking for a revision. I did want to ask though – is there a clearer (if longer) alternative implementation for this?

Reply:
Well, I think I brought this upon myself. I thought implementing `_*_to_glue_data` as iterator was clever because you process each loop one by one and don't have to read everything into memory all at once before processing the data. But as a result, when there is only one loop, you have to do weird things to it.

The cleaner way is to copy paste only that few lines from inside `_nddata_to_glue_data` to wherever this is called in production; for instance, you can throw away the unit handling stuff. But that means you are duplicating the parser code. But if that makes things easier to maintain going forward, so be it.

I am still hesitant to touch the actual `parsers.py` because that is meant for loading user data.